Automattic/wp-calypso
# Framework: Add CSS modules support

This pull request is a follow-up of #10366 that seeks to introduce [CSS modules](https://github.com/css-modules/css-modules) to Calypso. CSS modules ([here](https://css-tricks.com/css-modules-part-1-need/) is a good introduction) bring numerous advantages, the main one being that we no longer need to name classes [using a BEM-like syntax](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md) since all class and animation names are scoped locally by default. This means that it's now possible to use semantic and shorter names for our CSS classes.

So instead of importing a stylesheet in this massive [`_components.scss`](https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/_components.scss):

```javascript
@import 'components/vertical-nav/style';
```

And using a class name that must remain unique across Calypso:

```javascript
export default ( { children } ) => {
  return (
    <div className="vertical-nav">
      { children }
    </div>
  );
};
```

We can now just import the stylesheet from the component itself, as if it was a Javascript module:

```javascript
import style from './style';

export default ( { children } ) => {
  return (
    <div className={ style.content }>
      { children }
    </div>
  );
};
```

While still using plain good old CSS, but with shorter class names:

```css
.content {
  margin-top: 15px;
}
```

Webpack takes care of all the magic using three new [loaders](https://webpack.github.io/docs/loaders.html) and preprocesses SCSS files as you import them. To be more specific, the [CSS loader](https://github.com/webpack/css-loader) makes it possible to require CSS files from any component:

```javascript
import style from 'style.css';
```

The [SASS loader](https://github.com/jtangelder/sass-loader) unsurprisingly adds supports for `.scss` files:

```javascript
import style from 'style.scss';
```

Finally, the [Style loader](https://github.com/webpack/style-loader) adds the styles of a component to the `<head>` of the page. A prefix based on the path of the component is automatically injected in front of class names, thus preventing collisions:

```html
<html>
  <head>
    ...
    <style type="text/css">
      .client-components-vertical-nav-content {
        margin-top: 15px;
      }
    </style>
  </head>
  ...
```

#### Other benefits

* We can continue to use `@import` and `url()` in stylesheets since [they will be resolved by Webpack loaders](https://github.com/webpack/css-loader#options) (although all linked assets [must be relative](https://github.com/jtangelder/sass-loader#problems-with-url)).

* We can still target class names that have a global scope with the [`:global()` and `:global {}` constructs](https://github.com/webpack/css-loader#css-scope)

* We can still use composition, either by requiring different stylesheets from a component, or by using the [`composes` keyword](https://github.com/css-modules/css-modules#composition) from the stylesheets themselves (note that when composing multiple classes from different files the order of appliance is undefined).

* We only load CSS the first time a component is mounted.

* We can convert components to use CSS modules gradually.

* We can enable [source maps](https://github.com/jtangelder/sass-loader#source-maps) if we want to.


#### Testing instructions
 
1. Run `git checkout add/css-modules` and start your server, or open a [live branch](https://calypso.live/?branch=add/css-modules)
2. Log into a test account with at least a domain purchased
3. Open the [`Domains` page](http://calypso.localhost:3000/domains/manage)
4. Click the domain name to access the corresponding `Domain Settings` page
5. Check that navigation links are [correctly displayed](https://cloud.githubusercontent.com/assets/594356/21932008/b14d6824-d99e-11e6-864a-003fe1bcea9b.png) and can be clicked

#### Reviews
